In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

from imblearn.over_sampling import SMOTE
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.svm import SVC

In [2]:
#importing Austins feture engineered Bank Churn Dataframe

starter_df = pd.read_csv(Path("../Resources/BankChurnedit1.csv"))
starter_df

Unnamed: 0  Status  Gender  Education  Income  Card Type  \
0              0       1       0          1       3          0   
1              4       1       0          0       3          0   
2              8       1       0          0       3          0   
3             12       1       0          2       2          0   
4             16       1       0          3       2          0   
...          ...     ...     ...        ...     ...        ...   
4485       10113      -1       0          1       3          0   
4486       10115       1       0          0       1          0   
4487       10116       1       0          2       2          0   
4488       10120       1       0          1       3          0   
4489       10124      -1       1          1       0          0   

      Months_Inactive_12_mon  Total_Revolving_Bal  Avg_Utilization_Ratio  \
0                          1                  777                  0.061   
1                          1                    0                  0.000   
2                          2                 2517                  0.113   
3                          6                    0                  0.000   
4                          2                 2362                  0.078   
...                      ...                  ...                    ...   
4485                       3                 2517                  0.189   
4486                       3                 1558                  0.276   
4487                       2                 2241                  0.170   
4488                       2                 2109                  0.151   
4489                       3                    0                  0.000   

      MaritalDependentRatio  CreditUsage  AvgTransValue  TenureByAge  
0                      0.25         0.94          27.24         0.87  
1                      0.25         1.00          29.14         0.52  
2                      0.12         0.89          56.25         0.97  
3                      0.25         1.00          90.53         0.64  
4                      0.10         0.92          61.89         0.75  
...                     ...          ...            ...          ...  
4485                   0.50         0.81         120.22         1.33  
4486                   0.25         0.72         152.55         0.95  
4487                   0.08         0.83         137.09         0.78  
4488                   0.25         0.85         136.64         0.63  
4489                   0.50         1.00         171.52         0.82  

[4490 rows x 13 columns]

education only run all below for that

Best performing column compared to Status/Attrition 
Total Revolving Balance
Accuracy: 73.37%
              precision    recall  f1-score   support

          -1       0.33      0.55      0.41       192
           1       0.89      0.77      0.83       931

    accuracy                           0.73      1123
   macro avg       0.61      0.66      0.62      1123
weighted avg       0.80      0.73      0.76      1123


In [17]:
#X will be only Education
y = starter_df['Status']
X = starter_df['Education']

In [18]:
#Split features and target in to training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [19]:
#Use SMOTE to add synthetic data and balance our target feature value count
oversample = SMOTE()
X_train, y_train = oversample.fit_resample(X_train, y_train)

In [28]:
#Choose a model
#clf = SVC()
#clf = LogisticRegression()
clf = RandomForestClassifier(n_estimators= 2000)

In [29]:
#Fit
clf.fit(X_train, y_train)

RandomForestClassifier(n_estimators=2000)

In [30]:
#Predict
test_predictions = clf.predict(X_test)

In [31]:
# Evaluate the model's performance
accuracy = accuracy_score(test_predictions, y_test)
print("Accuracy: {:.2f}%".format(accuracy * 100))

Accuracy: 59.31%


In [32]:
test_matrix = confusion_matrix(y_test, test_predictions)
print(test_matrix)

[[ 81 111]
 [346 585]]


In [33]:
testing_report = classification_report(y_test, test_predictions)
print(testing_report)

              precision    recall  f1-score   support

          -1       0.19      0.42      0.26       192
           1       0.84      0.63      0.72       931

    accuracy                           0.59      1123
   macro avg       0.52      0.53      0.49      1123
weighted avg       0.73      0.59      0.64      1123



Education :
Accuracy: 59.31%
              precision    recall  f1-score   support

          -1       0.19      0.42      0.26       192
           1       0.84      0.63      0.72       931

    accuracy                           0.59      1123
   macro avg       0.52      0.53      0.49      1123
weighted avg       0.73      0.59      0.64      1123


In [ ]:
new test X will be Income category

In [34]:
#X will be only Income
y = starter_df['Status']
X = starter_df['Income']

In [35]:
#Split features and target in to training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [36]:
#Use SMOTE to add synthetic data and balance our target feature value count
oversample = SMOTE()
X_train, y_train = oversample.fit_resample(X_train, y_train)

In [37]:
#Choose a model
#clf = SVC()
#clf = LogisticRegression()
clf = RandomForestClassifier(n_estimators= 2000)

In [38]:
#Fit
clf.fit(X_train, y_train)

RandomForestClassifier(n_estimators=2000)

In [39]:
#Predict
test_predictions = clf.predict(X_test)

In [40]:
# Evaluate the model's performance
accuracy = accuracy_score(test_predictions, y_test)
print("Accuracy: {:.2f}%".format(accuracy * 100))

Accuracy: 61.26%


In [41]:
test_matrix = confusion_matrix(y_test, test_predictions)
print(test_matrix)

[[ 79 113]
 [322 609]]


In [42]:
testing_report = classification_report(y_test, test_predictions)
print(testing_report)

              precision    recall  f1-score   support

          -1       0.20      0.41      0.27       192
           1       0.84      0.65      0.74       931

    accuracy                           0.61      1123
   macro avg       0.52      0.53      0.50      1123
weighted avg       0.73      0.61      0.66      1123



Income
Accuracy: 61.26%
              precision    recall  f1-score   support

          -1       0.20      0.41      0.27       192
           1       0.84      0.65      0.74       931

    accuracy                           0.61      1123
   macro avg       0.52      0.53      0.50      1123
weighted avg       0.73      0.61      0.66      1123

In [44]:
#X will be only Card Type
y = starter_df['Status']
X = starter_df.drop(columns=['Status','Gender','Education','Income','Months_Inactive_12_mon','Total_Revolving_Bal','Avg_Utilization_Ratio','MaritalDependentRatio','CreditUsage','AvgTransValue','TenureByAge'])

In [45]:
#Split features and target in to training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [46]:
#Use SMOTE to add synthetic data and balance our target feature value count
oversample = SMOTE()
X_train, y_train = oversample.fit_resample(X_train, y_train)

In [47]:
#Choose a model
#clf = SVC()
#clf = LogisticRegression()
clf = RandomForestClassifier(n_estimators= 2000)

In [48]:
#Fit
clf.fit(X_train, y_train)

RandomForestClassifier(n_estimators=2000)

In [49]:
#Predict
test_predictions = clf.predict(X_test)

In [50]:
# Evaluate the model's performance
accuracy = accuracy_score(test_predictions, y_test)
print("Accuracy: {:.2f}%".format(accuracy * 100))

Accuracy: 56.81%


In [51]:
test_matrix = confusion_matrix(y_test, test_predictions)
print(test_matrix)

[[ 86 106]
 [379 552]]


In [52]:
testing_report = classification_report(y_test, test_predictions)
print(testing_report)

              precision    recall  f1-score   support

          -1       0.18      0.45      0.26       192
           1       0.84      0.59      0.69       931

    accuracy                           0.57      1123
   macro avg       0.51      0.52      0.48      1123
weighted avg       0.73      0.57      0.62      1123



Card Type
Accuracy: 56.81%
              precision    recall  f1-score   support

          -1       0.18      0.45      0.26       192
           1       0.84      0.59      0.69       931

    accuracy                           0.57      1123
   macro avg       0.51      0.52      0.48      1123
weighted avg       0.73      0.57      0.62      1123

In [53]:
#X will be only Months inactive 12 mon
y = starter_df['Status']
X = starter_df.drop(columns=['Status','Gender','Education','Income','Card Type','Total_Revolving_Bal','Avg_Utilization_Ratio','MaritalDependentRatio','CreditUsage','AvgTransValue','TenureByAge'])

In [54]:
#Split features and target in to training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [55]:
#Use SMOTE to add synthetic data and balance our target feature value count
oversample = SMOTE()
X_train, y_train = oversample.fit_resample(X_train, y_train)

In [56]:
#Choose a model
#clf = SVC()
#clf = LogisticRegression()
clf = RandomForestClassifier(n_estimators= 2000)

In [57]:
#Fit
clf.fit(X_train, y_train)

RandomForestClassifier(n_estimators=2000)

In [58]:
#Predict
test_predictions = clf.predict(X_test)

In [59]:
# Evaluate the model's performance
accuracy = accuracy_score(test_predictions, y_test)
print("Accuracy: {:.2f}%".format(accuracy * 100))

Accuracy: 63.22%


In [60]:
test_matrix = confusion_matrix(y_test, test_predictions)
print(test_matrix)

[[ 76 116]
 [297 634]]


In [61]:
testing_report = classification_report(y_test, test_predictions)
print(testing_report)

              precision    recall  f1-score   support

          -1       0.20      0.40      0.27       192
           1       0.85      0.68      0.75       931

    accuracy                           0.63      1123
   macro avg       0.52      0.54      0.51      1123
weighted avg       0.74      0.63      0.67      1123



Months inactive 12 mon
Accuracy: 63.22%
              precision    recall  f1-score   support

          -1       0.20      0.40      0.27       192
           1       0.85      0.68      0.75       931

    accuracy                           0.63      1123
   macro avg       0.52      0.54      0.51      1123
weighted avg       0.74      0.63      0.67      1123

In [12]:
#X will be only Total_Revolving_Bal
y = starter_df['Status']
X = starter_df.drop(columns=['Status','Gender','Education','Income','Card Type','Months_Inactive_12_mon','Avg_Utilization_Ratio','MaritalDependentRatio','CreditUsage','AvgTransValue','TenureByAge'])

In [13]:
#Split features and target in to training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [14]:
#Use SMOTE to add synthetic data and balance our target feature value count
oversample = SMOTE()
X_train, y_train = oversample.fit_resample(X_train, y_train)

In [15]:
#Choose a model
clf = SVC()
#clf = LogisticRegression()
#clf = RandomForestClassifier(n_estimators= 2000)

In [16]:
#Fit
clf.fit(X_train, y_train)

SVC()

In [17]:
#Predict
test_predictions = clf.predict(X_test)

In [18]:
# Evaluate the model's performance
accuracy = accuracy_score(test_predictions, y_test)
print("Accuracy: {:.2f}%".format(accuracy * 100))

Accuracy: 73.55%


In [19]:
test_matrix = confusion_matrix(y_test, test_predictions)
print(test_matrix)

[[132  60]
 [237 694]]


In [20]:
testing_report = classification_report(y_test, test_predictions)
print(testing_report)

              precision    recall  f1-score   support

          -1       0.36      0.69      0.47       192
           1       0.92      0.75      0.82       931

    accuracy                           0.74      1123
   macro avg       0.64      0.72      0.65      1123
weighted avg       0.82      0.74      0.76      1123



Total Revolving Balance
Accuracy: 73.37%
               precision    recall  f1-score   support

          -1       0.36      0.69      0.47       192
           1       0.92      0.75      0.82       931

    accuracy                           0.74      1123
   macro avg       0.64      0.72      0.65      1123
weighted avg       0.82      0.74      0.76      1123

In [30]:
#X will be only Avg_Utilization_Ratio
y = starter_df['Status']
X = starter_df.drop(columns=['Status','Gender','Education','Income','Card Type','Months_Inactive_12_mon','Total_Revolving_Bal','MaritalDependentRatio','CreditUsage','AvgTransValue','TenureByAge'])

In [31]:
#Split features and target in to training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [32]:
#Use SMOTE to add synthetic data and balance our target feature value count
oversample = SMOTE()
X_train, y_train = oversample.fit_resample(X_train, y_train)

In [33]:
#Choose a model
#clf = SVC()
#clf = LogisticRegression()
clf = RandomForestClassifier(n_estimators= 2000)

In [34]:
#Fit
clf.fit(X_train, y_train)

RandomForestClassifier(n_estimators=2000)

In [35]:
#Predict
test_predictions = clf.predict(X_test)

In [36]:
# Evaluate the model's performance
accuracy = accuracy_score(test_predictions, y_test)
print("Accuracy: {:.2f}%".format(accuracy * 100))

Accuracy: 67.14%


In [37]:
test_matrix = confusion_matrix(y_test, test_predictions)
print(test_matrix)

[[ 79 113]
 [256 675]]


In [38]:
testing_report = classification_report(y_test, test_predictions)
print(testing_report)

              precision    recall  f1-score   support

          -1       0.24      0.41      0.30       192
           1       0.86      0.73      0.79       931

    accuracy                           0.67      1123
   macro avg       0.55      0.57      0.54      1123
weighted avg       0.75      0.67      0.70      1123



RandomForestClassifier
Avg utilization ratio
Accuracy: 65.98%
              precision    recall  f1-score   support

          -1       0.23      0.42      0.30       192
           1       0.86      0.71      0.78       931

    accuracy                           0.66      1123
   macro avg       0.54      0.56      0.54      1123
weighted avg       0.75      0.66      0.69      1123


In [80]:
#X will be only MaritalDependentRatio 
y = starter_df['Status']
X = starter_df.drop(columns=['Status','Gender','Education','Income','Card Type','Months_Inactive_12_mon','Total_Revolving_Bal','Avg_Utilization_Ratio','CreditUsage','AvgTransValue','TenureByAge'])

In [81]:
#Split features and target in to training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [82]:
#Use SMOTE to add synthetic data and balance our target feature value count
oversample = SMOTE()
X_train, y_train = oversample.fit_resample(X_train, y_train)

In [83]:
#Choose a model
#clf = SVC()
#clf = LogisticRegression()
clf = RandomForestClassifier(n_estimators= 2000)

In [84]:
#Fit
clf.fit(X_train, y_train)

RandomForestClassifier(n_estimators=2000)

In [85]:
#Predict
test_predictions = clf.predict(X_test)

In [86]:
# Evaluate the model's performance
accuracy = accuracy_score(test_predictions, y_test)
print("Accuracy: {:.2f}%".format(accuracy * 100))

Accuracy: 71.95%


In [87]:
test_matrix = confusion_matrix(y_test, test_predictions)
print(test_matrix)

[[ 37 155]
 [160 771]]


In [88]:
testing_report = classification_report(y_test, test_predictions)
print(testing_report)

              precision    recall  f1-score   support

          -1       0.19      0.19      0.19       192
           1       0.83      0.83      0.83       931

    accuracy                           0.72      1123
   macro avg       0.51      0.51      0.51      1123
weighted avg       0.72      0.72      0.72      1123



Marital Dependent Ratio 
Accuracy: 71.95%
              precision    recall  f1-score   support

          -1       0.19      0.19      0.19       192
           1       0.83      0.83      0.83       931

    accuracy                           0.72      1123
   macro avg       0.51      0.51      0.51      1123
weighted avg       0.72      0.72      0.72      1123


In [66]:
#X will be only CreditUsage 
y = starter_df['Status']
X = starter_df.drop(columns=['Status','Gender','Education','Income','Card Type','Months_Inactive_12_mon','Total_Revolving_Bal','Avg_Utilization_Ratio','MaritalDependentRatio','AvgTransValue','TenureByAge'])

In [49]:
#Split features and target in to training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [50]:
#Use SMOTE to add synthetic data and balance our target feature value count
oversample = SMOTE()
X_train, y_train = oversample.fit_resample(X_train, y_train)

In [57]:
#Choose a model
#clf = SVC()
#clf = LogisticRegression()
clf = RandomForestClassifier(n_estimators= 2000)

In [58]:
#Fit
clf.fit(X_train, y_train)

RandomForestClassifier(n_estimators=2000)

In [59]:
#Predict
test_predictions = clf.predict(X_test)

In [60]:
# Evaluate the model's performance
accuracy = accuracy_score(test_predictions, y_test)
print("Accuracy: {:.2f}%".format(accuracy * 100))

Accuracy: 71.59%


In [63]:
test_matrix = confusion_matrix(y_test, test_predictions)
print(test_matrix)

[[ 69 123]
 [196 735]]


In [64]:
testing_report = classification_report(y_test, test_predictions)
print(testing_report)

              precision    recall  f1-score   support

          -1       0.26      0.36      0.30       192
           1       0.86      0.79      0.82       931

    accuracy                           0.72      1123
   macro avg       0.56      0.57      0.56      1123
weighted avg       0.75      0.72      0.73      1123



In [ ]:
CreditUsage
Accuracy: 71.59%
              precision    recall  f1-score   support

          -1       0.26      0.36      0.30       192
           1       0.86      0.79      0.82       931

    accuracy                           0.72      1123
   macro avg       0.56      0.57      0.56      1123
weighted avg       0.75      0.72      0.73      1123

In [99]:
#X will be only AvgTransValue
y = starter_df['Status']
X = starter_df.drop(columns=['Status','Gender','Education','Income','Card Type','Months_Inactive_12_mon','Total_Revolving_Bal','Avg_Utilization_Ratio','MaritalDependentRatio','CreditUsage','TenureByAge'])

In [100]:
#Split features and target in to training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [101]:
#Use SMOTE to add synthetic data and balance our target feature value count
oversample = SMOTE()
X_train, y_train = oversample.fit_resample(X_train, y_train)

In [102]:
#Choose a model
#clf = SVC()
#clf = LogisticRegression()
clf = RandomForestClassifier(n_estimators= 2000)

In [103]:
#Fit
clf.fit(X_train, y_train)

RandomForestClassifier(n_estimators=2000)

In [104]:
#Predict
test_predictions = clf.predict(X_test)

In [105]:
# Evaluate the model's performance
accuracy = accuracy_score(test_predictions, y_test)
print("Accuracy: {:.2f}%".format(accuracy * 100))

Accuracy: 69.10%


In [106]:
test_matrix = confusion_matrix(y_test, test_predictions)
print(test_matrix)

[[ 93  99]
 [248 683]]


In [107]:
testing_report = classification_report(y_test, test_predictions)
print(testing_report)

              precision    recall  f1-score   support

          -1       0.27      0.48      0.35       192
           1       0.87      0.73      0.80       931

    accuracy                           0.69      1123
   macro avg       0.57      0.61      0.57      1123
weighted avg       0.77      0.69      0.72      1123



In [ ]:
Avg Trans Value
Accuracy: 69.10%
              precision    recall  f1-score   support

          -1       0.27      0.48      0.35       192
           1       0.87      0.73      0.80       931

    accuracy                           0.69      1123
   macro avg       0.57      0.61      0.57      1123
weighted avg       0.77      0.69      0.72      1123

In [108]:
#X will be only TenureByAge
y = starter_df['Status']
X = starter_df.drop(columns=['Status','Gender','Education','Income','Card Type','Months_Inactive_12_mon','Total_Revolving_Bal','Avg_Utilization_Ratio','MaritalDependentRatio','CreditUsage','AvgTransValue'])

In [109]:
#Split features and target in to training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [110]:
#Use SMOTE to add synthetic data and balance our target feature value count
oversample = SMOTE()
X_train, y_train = oversample.fit_resample(X_train, y_train)

In [111]:
#Choose a model
#clf = SVC()
#clf = LogisticRegression()
clf = RandomForestClassifier(n_estimators= 2000)

In [112]:
#Fit
clf.fit(X_train, y_train)

RandomForestClassifier(n_estimators=2000)

In [113]:
#Predict
test_predictions = clf.predict(X_test)

In [114]:
# Evaluate the model's performance
accuracy = accuracy_score(test_predictions, y_test)
print("Accuracy: {:.2f}%".format(accuracy * 100))

Accuracy: 68.66%


In [115]:
test_matrix = confusion_matrix(y_test, test_predictions)
print(test_matrix)

[[ 43 149]
 [203 728]]


In [116]:
testing_report = classification_report(y_test, test_predictions)
print(testing_report)

              precision    recall  f1-score   support

          -1       0.17      0.22      0.20       192
           1       0.83      0.78      0.81       931

    accuracy                           0.69      1123
   macro avg       0.50      0.50      0.50      1123
weighted avg       0.72      0.69      0.70      1123



Tenure ByAge
Accuracy: 68.66%
              precision    recall  f1-score   support

          -1       0.17      0.22      0.20       192
           1       0.83      0.78      0.81       931

    accuracy                           0.69      1123
   macro avg       0.50      0.50      0.50      1123
weighted avg       0.72      0.69      0.70      1123
